Sources/SwiftProtobuf/ProtobufTextDecoding.swift

```diff
@@ -0,0 +1,1229 @@
+            //     more than 20 digits wihout overflow
+            //   = Value must be integral, so a negative exponent
+            //     can't be greater than number of digits
+            // The limit here is deliberately sloppy, it is only intended
+            // to avoid painful abuse cases (e.g., 1e1000000000 will be
+            // quickly dropped without trying to build a an array
+            // of a billion characters).
+            if let exp = Int(expString), exp + digits < 20 && exp > -digits {
+                // Fold fractional digits into exponent
+                var adjustment = exp - (fractionalDigits ?? 0)
+                fractionalDigits = 0
+                // Adjust digit string to account for exponent
+                while adjustment > 0 {
+                    number.append("0")
+                    adjustment -= 1
+                }
+                while adjustment < 0 {
+                    if number.isEmpty || number[number.count - 1] != "0" {
+                        return nil
+                    }
+                    number.remove(at: number.count - 1)
+                    adjustment += 1
+                }
+            } else {
+                // Error if exponent is malformed or out of range
+                return nil
+            }
+        default:
+            return nil
+        }
+    }
+    if number.isEmpty {
+        return nil
+    }
+    // Allow 7.000 and 1.23000e2 by trimming fractional zero digits
+    if let f = fractionalDigits {
+        var fractionalDigits = f
+        while fractionalDigits > 0 && !number.isEmpty && number[number.count - 1] == "0" {
+            number.remove(at: number.count - 1)
+            fractionalDigits -= 1
+        }
+        if fractionalDigits > 0 {
+            return nil
+        }
+    }
+    return number
+}
+
+private func decodeBytes(_ s: String) -> Data? {
+    var out = [UInt8]()
+    let digits = s.utf8
+    var n = 0
+    var bits = 0
+    for (i, digit) in digits.enumerated() {
+        n <<= 6
+        switch digit {
+        case 65...90: n |= Int(digit - 65); bits += 6
+        case 97...122: n |= Int(digit - 97 + 26); bits += 6
+        case 48...57: n |= Int(digit - 48 + 52); bits += 6
+        case 43: n |= 62; bits += 6
+        case 47: n |= 63; bits += 6
+        case 61: n |= 0
+        default:
+            return nil
+        }
+        if i % 4 == 3 {
+            out.append(UInt8(truncatingBitPattern: n >> 16))
+            if bits >= 16 {
+                out.append(UInt8(truncatingBitPattern: n >> 8))
+                if bits >= 24 {
+                    out.append(UInt8(truncatingBitPattern: n))
+                }
+            }
+            bits = 0
+        }
+    }
+    if bits != 0 {
+        return nil
+    }
+    return Data(bytes: out)
+}
+
+
+public enum ProtobufTextToken: Equatable, ProtobufFieldDecoder {
+    case colon
+    case comma
+    case beginObject
+    case endObject
+    case beginArray
+    case endArray
```

Comment:
Repeated fields can be repeated in text format, but there is also a shorthand bracket syntax: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L527